hbase-common/src/main/java/org/apache/hadoop/hbase/CellComparatorImpl.java

```diff
@@ -377,6 +378,27 @@ private static int compareRows(byte[] left, int loffset, int llength, byte[] rig
       return result;
     }
 
+    @Override
+    public int compareRows(ByteBuffer row, Cell cell) {
+      byte [] array;
+      int offset;
+      int len = row.remaining();
+      if (row.hasArray()) {
+        array = row.array();
+        offset = row.position() + row.arrayOffset();
+      } else {
+        // This is awful, we copy the row array if offheap just so we can do a compare.
+        // We do this elsewhere too when Cell is backed by an offheap ByteBuffer.
+        // Needs fixing. TODO.
```

Comment:
No.

Since making this comment, I see that we do this whenever the BB is offheap.  Let me make an issue. This issue adds a method to CellComparator that takes a row held in a ByteBuffer. We need more of this with methods in Comparator that can stride through a BB by index so we don't have to copy on heap to compare. The hard part is an implementation that does not insist on two compare methods -- one for array and another for BB.   Would be good if could have one compare only.

Reply:
Are you sure on this STack. All the impl of the BB vs Array (vice versa) or BB vs BB all happens inside the BBUtils method. There we are clearly ensuring that we either call getXXxOffset() or getXXXPosition() so that we avoide copying onheap. Only when a user tries to do getRowArray on a BB backed cell we try to copy the row part and then do the comparsion. In that case he should also be using getRowOfset() only which would be 0/.

Reply:
> Only when a user tries to do getRowArray on a BB backed cell we try to copy the row part and then do the comparsion. In that case he should also be using getRowOfset() only which would be

I should have been more specific. Lets make it so no need to copy-to-compare ever.

Reply:
Made the comment more mild. Added pointer to BBUtils, that we do here what it does. Added note to fix ever having to copy on heap.